bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/ScalingFontRegistry.java

```diff
@@ -0,0 +1,181 @@
+		systemFont = Font.win32_new(display, systemFontHandle, zoomFactor);
+		container.addScaledFont(zoomFactor, systemFont);
+		return systemFont;
+	}
+
+	private ScaledFontContainer getOrCreateBaseSystemFontContainer(Display display) {
+		ScaledFontContainer systemFontContainer = fontKeyMap.get(KEY_SYSTEM_FONTS);
+		if (systemFontContainer == null) {
+			int targetZoomFactor = display.getPrimaryMonitor().getZoom();
+			long systemFontHandle = createSystemFont(display, targetZoomFactor);
+			Font systemFont = Font.win32_new(display, systemFontHandle);
+			systemFontContainer = new ScaledFontContainer(systemFont);
+			fontHandleMap.put(systemFont.handle, systemFontContainer);
+			fontKeyMap.put(KEY_SYSTEM_FONTS, systemFontContainer);
+		}
+		return systemFontContainer;
+	}
+
+	private long createSystemFont(Display display, int targetZoomFactor) {
+		long hFont = 0;
+		NONCLIENTMETRICS info = new NONCLIENTMETRICS();
+		info.cbSize = NONCLIENTMETRICS.sizeof;
+		if (fetchSystemParametersInfo(info, targetZoomFactor)) {
+			LOGFONT logFont = info.lfMessageFont;
+			hFont = OS.CreateFontIndirect(logFont);
+		}
+		if (hFont == 0)
+			hFont = OS.GetStockObject(OS.DEFAULT_GUI_FONT);
+		if (hFont == 0)
+			hFont = OS.GetStockObject(OS.SYSTEM_FONT);
+		return hFont;
+	}
+
+	private static boolean fetchSystemParametersInfo(NONCLIENTMETRICS info, int targetZoomFactor) {
+		if (OS.WIN32_BUILD >= OS.WIN32_BUILD_WIN10_1607) {
+			return OS.SystemParametersInfoForDpi(OS.SPI_GETNONCLIENTMETRICS, NONCLIENTMETRICS.sizeof, info, 0,
+					DPIUtil.mapZoomToDPI(targetZoomFactor));
+		} else {
+			return OS.SystemParametersInfo(OS.SPI_GETNONCLIENTMETRICS, 0, info, 0);
+		}
+	}
+
+	@Override
+	public Font getFont(FontData fontData, int zoomFactor) {
+		if (!DPIUtil.autoScaleOnRuntime) {
+			return null;
+		}
+		ScaledFontContainer container;
+		if (fontKeyMap.containsKey(fontData)) {
+			container = fontKeyMap.get(fontData);
+		} else {
+			int calculatedZoomFactor = computeZoomFactor(fontData);
+			Font newFont = Font.win32_new(display, fontData, calculatedZoomFactor);
+			container = new ScaledFontContainer(newFont);
+			fontHandleMap.put(newFont.handle, container);
+			fontKeyMap.put(fontData, container);
+		}
+		return getOrCreateFont(container, zoomFactor);
+	}
+
+	@Override
+	public void dispose() {
+
+	}
+
+	private Font getOrCreateFont(ScaledFontContainer container, int zoomFactor) {
+		Font scaledFont = container.getScaledFont(zoomFactor);
+		if (scaledFont == null) {
+			scaledFont = container.scaleFont(zoomFactor);
+			fontHandleMap.put(scaledFont.handle, container);
+			fontKeyMap.put(scaledFont.getFontData()[0], container);
+		}
+		return scaledFont;
+	}
+
+	private int computeZoomFactor(FontData fontData) {
+		long hDC = display.internal_new_GC(null);
+		int pixelsAtPrimaryMonitorZoom = -(int) (0.5f + (fontData.height * OS.GetDeviceCaps(hDC, OS.LOGPIXELSY) / 72f));
+		display.internal_dispose_GC(hDC, null);
```

Comment:
This looks like a direct code copy of what is implemented in `computePixels()`. In addition, there is quite some magic in the formula (in particular the 72), which should be explained to make it comprehensible.

Reply:
My fault. I was confused, why I couldn't used `computePixels(...)` due to visibility reasons, but could just do so by casting the `Display `to `Device`